### ISSUE TITLE
Have you tried to train on combing multiple datasets such as vehicleID and veriwild?

### ISSUE THREAD

**Issue by HouBiaoLiu** [JDAI-CV/fast-reid#163]
We can get an excellent model training separately, but how to get a more general model supporting multiple datasets?

**Comment by L1aoXingyu** [comment]
It's a research problem not solved yet.
Maybe you can just combine these two datasets to train, but there will be little drop in these two datasets.

**Comment by HouBiaoLiu** [comment]
[07/17 15:21:09] fastreid.evaluation.evaluator INFO: Total inference pure compute time: 0:00:16 (1.083325 s / img per device)
[07/17 15:21:32] fastreid.engine.defaults INFO: Evaluation results for LargeVehicleID in csv format:
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: Rank-1
[07/17 15:21:32] fastreid.evaluation.testing INFO: 77.7%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: Rank-5
[07/17 15:21:32] fastreid.evaluation.testing INFO: 90.3%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: Rank-10
[07/17 15:21:32] fastreid.evaluation.testing INFO: 95.0%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: mAP
[07/17 15:21:32] fastreid.evaluation.testing INFO: 83.3%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: mINP
[07/17 15:21:32] fastreid.evaluation.testing INFO: 83.3%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: TPR@FPR=0.0001
[07/17 15:21:32] fastreid.evaluation.testing INFO: 8.8%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: TPR@FPR=0.001
[07/17 15:21:32] fastreid.evaluation.testing INFO: 48.8%
[07/17 15:21:32] fastreid.evaluation.testing INFO: Task: TPR@FPR=0.01
[07/17 15:21:32] fastreid.evaluation.testing INFO: 98.4%
[07/17 15:21:32] fastreid.engine.defaults INFO: prepare test set
[07/17 15:21:35] fastreid.data.datasets.bases INFO: => Loaded SmallVeRiWild
[07/17 15:21:35] fastreid.data.datasets.bases INFO:   ----------------------------------------
[07/17 15:21:35] fastreid.data.datasets.bases INFO:   subset   | # ids | # images | # cameras
[07/17 15:21:35] fastreid.data.datasets.bases INFO:   ----------------------------------------
[07/17 15:21:35] fastreid.data.datasets.bases INFO:   query    |  3000 |     3000 |       105
[07/17 15:21:35] fastreid.data.datasets.bases INFO:   gallery  |  3000 |    38861 |       146
[07/17 15:21:35] fastreid.data.datasets.bases INFO:   ----------------------------------------
[07/17 15:21:35] fastreid.evaluation.evaluator INFO: Start inference on 41861 images
[07/17 15:23:15] fastreid.evaluation.evaluator INFO: Inference done 11/41. 0.9825 s / img. ETA=0:00:32
[07/17 15:24:17] fastreid.evaluation.evaluator INFO: Inference done 21/41. 1.0777 s / img. ETA=0:01:26
[07/17 15:24:49] fastreid.evaluation.evaluator INFO: Inference done 37/41. 1.1272 s / img. ETA=0:00:12
[07/17 15:24:59] fastreid.evaluation.evaluator INFO: Total inference time: 0:01:50.250082 (3.062502 s / img per device)
[07/17 15:24:59] fastreid.evaluation.evaluator INFO: Total inference pure compute time: 0:00:41 (1.151221 s / img per device)
[07/17 15:26:12] fastreid.engine.defaults INFO: Evaluation results for SmallVeRiWild in csv format:
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: Rank-1
[07/17 15:26:12] fastreid.evaluation.testing INFO: 93.6%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: Rank-5
[07/17 15:26:12] fastreid.evaluation.testing INFO: 97.3%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: Rank-10
[07/17 15:26:12] fastreid.evaluation.testing INFO: 98.6%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: mAP
[07/17 15:26:12] fastreid.evaluation.testing INFO: 74.1%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: mINP
[07/17 15:26:12] fastreid.evaluation.testing INFO: 46.5%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: TPR@FPR=0.0001
[07/17 15:26:12] fastreid.evaluation.testing INFO: 8.3%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: TPR@FPR=0.001
[07/17 15:26:12] fastreid.evaluation.testing INFO: 57.3%
[07/17 15:26:12] fastreid.evaluation.testing INFO: Task: TPR@FPR=0.01
[07/17 15:26:12] fastreid.evaluation.testing INFO: 99.3%


From the log.txt we can find out the map is much lower than the seperate trained SmallVeRiWild  result, but Rank-1 is near the seperate trained SmallVeRiWild  result

**Comment by L1aoXingyu** [comment]
It's an open question and maybe you can check this paper to get some help.

https://arxiv.org/pdf/2003.07618.pdf

**Comment by L1aoXingyu** [comment]
We will conduct some experiments on this question. Maybe this weekend, we can get some conclusions.

**Comment by L1aoXingyu** [comment]
These days, we just conduct some experiments and found the number of domains will boost the performance. 

If the performance dropped when you combine two domains, you can combine more domains and the performance will get better.

![image](https://user-images.githubusercontent.com/19245504/90222644-12ec3f80-de3f-11ea-959c-9b47bd47d082.png)

The training set is `Market1501+CUHK03+MSMT17` and `DukeMTMC` is directly transferred results.

**Comment by TCHeish** [comment]
> These days, we just conduct some experiments and found the number of domains will boost the performance.
> 
> If the performance dropped when you combine two domains, you can combine more domains and the performance will get better.
> 
> ![image](https://user-images.githubusercontent.com/19245504/90222644-12ec3f80-de3f-11ea-959c-9b47bd47d082.png)
> 
> The training set is `Market1501+CUHK03+MSMT17` and `DukeMTMC` is directly transferred results.

Hi, 

1. Which  config did you used ?
2. Have you try using GAN generated data ? How's the results?
3. To address cross-domain problem, have you try  r50-ibn-b backbone ? according to the paper, this backbone performs better in cross-domain.

Thanks!

**Comment by HouBiaoLiu** [comment]
@L1aoXingyu  Ok, thanks for your reply! I have read your suggested paper. As we see, it adopts the osnet, replaces global average pooling operation with global depthwise convolution  and uses AM-softmax loss. So what is your experiment setup? The other question is when we want to improve the performance on  veriwild, we should introduce the camera-aware model and direction-aware model?

**Comment by HouBiaoLiu** [comment]
![image](https://user-images.githubusercontent.com/13248587/91028113-93752200-e62f-11ea-89a5-983d5dbd76e9.png)
Compared to your result, it shows better?

**Comment by L1aoXingyu** [comment]
> 1. Which  config did you used ?
> 2. Have you try using GAN generated data ? How's the results?
> 3. To address cross-domain problem, have you try  r50-ibn-b backbone ? according to the paper, this backbone performs better in cross-domain.
> 
> Thanks!

1. I just use `circleSoftmax` and `tripletloss`, use SGD optimiser with lr=0.01, training 60 epochs with cosine lr scheduler.
2. I tried GAN generated data before, its gain is very weak or even has no gain.
3. I don't try R50-ibn-b, maybe you can try it by yourself and share it with us. Or you can wait me to have time trying it.

@HouBiaoLiu The result in the paper is **combineall** training, and if I combine all the datasets and add my private data, I use R50-ibn-a can get result below.

![image](https://user-images.githubusercontent.com/19245504/91043931-0b9b1200-e647-11ea-9cbb-cfb3ab651eff.png)

**Comment by HouBiaoLiu** [comment]
Your result is very appealing on person reid experiment! As in the vehicle reid, the veri-wild datasets have many angles, when combined with vehicleID to train, we should apply affine transform

**Comment by L1aoXingyu** [comment]
Do you mean you use keypoint to apply the affine transform to train vehicle re-id?

**Comment by HouBiaoLiu** [comment]
When we combine both vehicleID and veri-wild to train, we should use data augmentantion such as random affine transform

**Comment by HouBiaoLiu** [comment]
@L1aoXingyu   Today I check the code in the common.py,  when I use multiple datasets to train, and if dataset A and dataset B have the same ids, the relabel operation would combine the id, the get_pids function seems no effect. 

# encoding: utf-8
"""
@author:  liaoxingyu
@contact: sherlockliao01@gmail.com
"""

import torch
from torch.utils.data import Dataset

from .data_utils import read_image


class CommDataset(Dataset):
    """Image Person ReID Dataset"""

    def __init__(self, img_items, transform=None, relabel=True):
        self.transform = transform
        self.relabel = relabel

        self.pid_dict = {}
        if self.relabel:
            self.img_items = []
            pids = set()
            for i, item in enumerate(img_items):
                pid = self.get_pids(item[0], item[1])
                self.img_items.append((item[0], pid, item[2]))  # replace pid
                pids.add(pid)
            self.pids = pids
            print (len(self.pids))
            self.pid_dict = dict([(p, i) for i, p in enumerate(self.pids)])
        else:
            self.img_items = img_items

    def __len__(self):
        return len(self.img_items)

    def __getitem__(self, index):
        img_path, pid, camid = self.img_items[index]
        img = read_image(img_path)
        if self.transform is not None: img = self.transform(img)

        if self.relabel: pid = self.pid_dict[pid]

        return {
            'images': img,
            'targets': pid,
            'camid': camid,
            'img_path': img_path
        }

    @staticmethod
    def get_pids(file_path, pid):
        """ Suitable for muilti-dataset training """
        if 'cuhk03' in file_path: prefix = 'cuhk'
        else:                     prefix = file_path.split('/')[1]

        return prefix + '_' + str(pid)

    def update_pid_dict(self, pid_dict):
        self.pid_dict = pid_dict

**Comment by HouBiaoLiu** [comment]
the train is normal, the total id nums is set in the configure file, may be this bug causes that the map is much lower than the seperate trained SmallVeRiWild result?

**Comment by HouBiaoLiu** [comment]
@L1aoXingyu  when I use vim to open this file events.out.tfevents.1594058219.ubuntu.19110.0, I found it shows error code! So what is the ture way to open this file?

**Comment by L1aoXingyu** [comment]
> when I use vim to open this file events.out.tfevents.1594058219.ubuntu.19110.0, I found it shows error code! So what is the ture way to open this file?

This should be opened with `tensorboard` with command `tensorboard --logdir dir-path/to/file`

Your `CommDataset` is too old, you should check the latest version and there are many updates.

**Comment by HouBiaoLiu** [comment]
Ok, thanks for your reply! I use tensorboard  to oberserve the rank curve and cls_accuracy curve.
![image](https://user-images.githubusercontent.com/13248587/91536817-e3adf600-e947-11ea-9cdd-e12332af4762.png)
![image](https://user-images.githubusercontent.com/13248587/91536854-f6c0c600-e947-11ea-9665-8827cfa9f337.png)
Another question is how to use this project to train fine-grained classification? Specially some classes have only four or less images in them.

**Comment by L1aoXingyu** [comment]
> Another question is how to use this project to train fine-grained classification? Specially some classes have only four or less images in them also for the reid task.

This might be called long-tail classification, not fine-grained classification. Actually it's very easy to make fastreid to support classification problem, you only need to define a custom classifier head. We will support it very soon.

**Comment by howieraem** [comment]
> > 1. Which  config did you used ?
> > 2. Have you try using GAN generated data ? How's the results?
> > 3. To address cross-domain problem, have you try  r50-ibn-b backbone ? according to the paper, this backbone performs better in cross-domain.
> > 
> > Thanks!
> 
> 1. I just use `circleSoftmax` and `tripletloss`, use SGD optimiser with lr=0.01, training 60 epochs with cosine lr scheduler.
> 2. I tried GAN generated data before, its gain is very weak or even has no gain.
> 3. I don't try R50-ibn-b, maybe you can try it by yourself and share it with us. Or you can wait me to have time trying it.
> 
> @HouBiaoLiu The result in the paper is **combineall** training, and if I combine all the datasets and add my private data, R50-ibn-a can get results below.
> 
> ![image](https://user-images.githubusercontent.com/19245504/91043931-0b9b1200-e647-11ea-9cbb-cfb3ab651eff.png)

Hello, many thanks for your work and reporting your results! I was trying to reproduce with the following config. I did't have private data and I might have different settings on backbone, CJ and iteration. However, I could only get ~63% rank1-accuracy on Duke dataset. Could you please point out what could possibly be wrong below?
```
CUDNN_BENCHMARK: true
DATALOADER:
  NAIVE_WAY: true
  NUM_INSTANCE: 16
  NUM_WORKERS: 8
  PK_SAMPLER: true
DATASETS:
  COMBINEALL: true
  NAMES:
  - Market1501
  - MSMT17
  TESTS:
  - DukeMTMC
INPUT:
  CJ:
    BRIGHTNESS:
    - 0.95
    - 1.15
    CONTRAST:
    - 0.85
    - 1.1
    ENABLED: true
    HUE: 0.1
    PROB: 0.8
    SATURATION:
    - 0.9
    - 1.05
  DO_AUGMIX: false
  DO_AUTOAUG: true
  DO_FLIP: true
  DO_PAD: true
  FLIP_PROB: 0.5
  PADDING: 10
  PADDING_MODE: constant
  REA:
    ENABLED: true
    MEAN:
    - 123.675
    - 116.28
    - 103.53
    PROB: 0.5
  RPT:
    ENABLED: false
    PROB: 0.5
  SIZE_TEST:
  - 384
  - 128
  SIZE_TRAIN:
  - 384
  - 128
MODEL:
  BACKBONE:
    DEPTH: 50x
    FEAT_DIM: 2048
    LAST_STRIDE: 1
    NAME: build_resnet_backbone
    NORM: syncBN
    PRETRAIN: true
    PRETRAIN_PATH: resnet50-ibn.pth
    WITH_IBN: true
    WITH_NL: false
    WITH_SE: false
  DEVICE: cuda
  FREEZE_LAYERS:
  - backbone
  - b1
  - b2
  - b3
  HEADS:
    CLS_LAYER: circleSoftmax
    EMBEDDING_DIM: 256
    MARGIN: 0.35
    NAME: EmbeddingHead
    NECK_FEAT: after
    NORM: syncBN
    NUM_CLASSES: 0
    POOL_LAYER: gempoolP
    SCALE: 64
    WITH_BNNECK: true
  LOSSES:
    CE:
      ALPHA: 0.2
      EPSILON: 0.1
      SCALE: 1.0
    CIRCLE:
      ALPHA: 128
      MARGIN: 0.25
      SCALE: 1.0
    FL:
      ALPHA: 0.25
      GAMMA: 2
      SCALE: 1.0
    NAME:
    - CrossEntropyLoss
    - TripletLoss
    TRI:
      HARD_MINING: true
      MARGIN: 0.0
      NORM_FEAT: false
      SCALE: 1.0
  META_ARCHITECTURE: MGN
  PIXEL_MEAN:
  - 123.675
  - 116.28
  - 103.53
  PIXEL_STD:
  - 58.395
  - 57.120000000000005
  - 57.375
  WEIGHTS: ''
OUTPUT_DIR: logs/aug
SOLVER:
  AMP_ENABLED: false
  BASE_LR: 0.00035
  BIAS_LR_FACTOR: 1.0
  CHECKPOINT_PERIOD: 5
  DELAY_ITERS: 30
  ETA_MIN_LR: 7.7e-07
  FREEZE_ITERS: 10
  GAMMA: 0.1
  HEADS_LR_FACTOR: 1.0
  IMS_PER_BATCH: 64
  MAX_ITER: 75
  MOMENTUM: 0.9
  OPT: Adam
  SCHED: WarmupCosineAnnealingLR
  STEPS:
  - 40
  - 90
  SWA:
    ENABLED: false
    ETA_MIN_LR: 3.5e-06
    ITER: 10
    LR_FACTOR: 10.0
    LR_SCHED: false
    PERIOD: 2
  WARMUP_FACTOR: 0.01
  WARMUP_ITERS: 10
  WARMUP_METHOD: linear
  WEIGHT_DECAY: 0.0005
  WEIGHT_DECAY_BIAS: 0.0005
TEST:
  AQE:
    ALPHA: 3.0
    ENABLED: false
    QE_K: 5
    QE_TIME: 1
  EVAL_PERIOD: 5
  IMS_PER_BATCH: 128
  METRIC: cosine
  PRECISE_BN:
    DATASET: Market1501
    ENABLED: false
    NUM_ITER: 300
  RERANK:
    ENABLED: false
    K1: 20
    K2: 6
    LAMBDA: 0.3
  ROC_ENABLED: false
```

**Comment by L1aoXingyu** [comment]
It seems you were using MGN arch. Actually, you want to boost the performance on Duke, you'd better add more datasets.

**Comment by howieraem** [comment]
> It seems you were using MGN arch. Actually, you want to boost the performance on Duke, you'd better add more datasets.

Thanks for your suggestion. By the way, have you tested with the extra 500K distractors in Market1501 (Market-1501+500k as mentioned here [https://zhuanlan.zhihu.com/p/38329352](url)? If so, which datasets would you recommend using for training?

**Comment by KRYSTALLJY** [comment]
> omains, you can combine more domains and the

Hi, can you share your training parameters. I combine Market, DukeMTMC, CUHK, and PRID, and get the rank-1 73%

**Comment by edwardnguyen1705** [comment]
any update on this issue?

**Comment by qq995431104** [comment]
> 

请问,如何将多个数据集的评估结果在一个表中显示?我对多个数据集评估, 是分开的.

**Comment by qq995431104** [comment]
> These days, we just conduct some experiments and found the number of domains will boost the performance.
> 
> If the performance dropped when you combine two domains, you can combine more domains and the performance will get better.
> 
> ![image](https://user-images.githubusercontent.com/19245504/90222644-12ec3f80-de3f-11ea-959c-9b47bd47d082.png)
> 
> The training set is `Market1501+CUHK03+MSMT17` and `DukeMTMC` is directly transferred results.

请问,如何将多个数据集的评估结果在一个表中显示?我对多个数据集评估, 是分开的.
